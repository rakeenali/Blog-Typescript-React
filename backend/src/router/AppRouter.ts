import express from "express";

export default class AppRouter {
  static instance: express.Router;

  static getRouter(): express.Router {
    if (!AppRouter.instance) {
      AppRouter.instance = express.Router();
    }

    return AppRouter.instance;
  }
}
