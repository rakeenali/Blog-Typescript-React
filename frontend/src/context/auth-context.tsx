import * as React from "react";
import { navigate } from "@reach/router";

import { Dispatch, ContextProps } from "../types/types";

export interface IAuthState {
  isAuthenticated: boolean;
  token: string;
}

type Action = { type: "LOGOUT" } | { type: "LOGIN"; payload: string };

export const AuthStateContext = React.createContext<IAuthState | null>(null);
export const AuthDispatchContext = React.createContext<Dispatch<Action> | null>(
  null
);

const authReducer = (state: IAuthState, action: Action) => {
  switch (action.type) {
    case "LOGIN": {
      return { ...state, isAuthenticated: true, token: action.payload };
    }
    case "LOGOUT": {
      return { ...INITIALSTATE };
    }
    default: {
      throw new Error(`Action type is not found in user context`);
    }
  }
};

const INITIALSTATE: IAuthState = {
  isAuthenticated: false,
  token: ""
};

export const AuthProvider = ({ children }: ContextProps): JSX.Element => {
  const [state, dispatch] = React.useReducer(authReducer, INITIALSTATE);

  return (
    <AuthStateContext.Provider value={state}>
      <AuthDispatchContext.Provider value={dispatch}>
        {children}
      </AuthDispatchContext.Provider>
    </AuthStateContext.Provider>
  );
};

export function useAuthState(): IAuthState {
  const Context = React.useContext(AuthStateContext);

  if (!Context) {
    throw new Error(`useAuthState must be used iniside auth provider`);
  }

  return Context;
}

export function useAuthDispatch(): Dispatch<Action> {
  const Context = React.useContext(AuthDispatchContext);

  if (!Context) {
    throw new Error(`useAuthDispatch must be used iniside auth provider`);
  }

  return Context;
}

export const setUserAction = (
  token: string,
  dispatch: Dispatch<Action>
): void => {
  navigate("/profile");
  localStorage.setItem("blog:key", JSON.stringify(token));
  dispatch({ type: "LOGIN", payload: token });
};

export const getUserAction = (dispatch: Dispatch<Action>): void => {
  const rawToken = localStorage.getItem("blog:key");
  if (rawToken) {
    const token: string = JSON.parse(rawToken);
    dispatch({ type: "LOGIN", payload: token });
  }
};

export const logout = (dispatch: Dispatch<Action>): void => {
  localStorage.removeItem("blog:key");
  navigate("/login");
  dispatch({ type: "LOGOUT" });
};
