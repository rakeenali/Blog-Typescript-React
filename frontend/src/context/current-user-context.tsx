import * as React from "react";
import { CurrentUserResponse as CurrentUserState } from "../types/inteface";
import { useAuthState } from "./auth-context";
import axios, { AxiosResponse } from "axios";
import { Dispatch, ContextProps } from "../types/types";

type Action = { type: "SET_USER"; payload: CurrentUserState };

export const CurrentUserStateContext = React.createContext<CurrentUserState | null>(
  null
);
export const CurrentUserDispatchContext = React.createContext<Dispatch<
  Action
> | null>(null);

const currentUserReducer = (state: CurrentUserState, action: Action) => {
  switch (action.type) {
    case "SET_USER": {
      return {
        ...state,
        ...action.payload
      };
    }
    default: {
      throw new Error(`Action type is not found in user context`);
    }
  }
};

const INITIALSTATE: CurrentUserState = {
  userId: null,
  name: null,
  username: null,
  usersProfile: null,
  profile: null,
  blogs: []
};

export const CurrentUserProvider = ({
  children
}: ContextProps): JSX.Element => {
  const [state, dispatch] = React.useReducer(currentUserReducer, INITIALSTATE);
  const authState = useAuthState();

  React.useLayoutEffect(() => {
    if (authState.isAuthenticated) {
      axios
        .get("http://localhost:5000/profile/current", {
          headers: { authorization: authState.token }
        })
        .then((res: AxiosResponse<CurrentUserState>) => {
          dispatch({ type: "SET_USER", payload: res.data });
        });
    }
  }, [authState]);

  return (
    <CurrentUserStateContext.Provider value={state}>
      <CurrentUserDispatchContext.Provider value={dispatch}>
        {children}
      </CurrentUserDispatchContext.Provider>
    </CurrentUserStateContext.Provider>
  );
};

export function useCurrentUserState(): CurrentUserState {
  const Context = React.useContext(CurrentUserStateContext);

  if (!Context) {
    throw new Error(
      `useCurrentUserState must be used iniside CurreuntUser provider`
    );
  }

  return Context;
}

export function useCurrentUserDispatch(): Dispatch<Action> {
  const Context = React.useContext(CurrentUserDispatchContext);

  if (!Context) {
    throw new Error(
      `useCurrentUserDispatch must be used iniside CurreuntUser provider`
    );
  }

  return Context;
}

export const getProfile = (token: string, dispatch: Dispatch<Action>): void => {
  axios
    .get("http://localhost:5000/profile/current", {
      headers: { authorization: token }
    })
    .then((res: AxiosResponse<CurrentUserState>) => {
      dispatch({ type: "SET_USER", payload: res.data });
    });
};
