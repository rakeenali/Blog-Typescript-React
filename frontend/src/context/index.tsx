import React from "react";

import { AuthProvider } from "./auth-context";

type AppProps = { children: React.ReactNode };

export default function AppProvider({ children }: AppProps): JSX.Element {
  return <AuthProvider>{children}</AuthProvider>;
}
