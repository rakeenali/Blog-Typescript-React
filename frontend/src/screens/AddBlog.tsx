import React from "react";
import styled from "styled-components";
import axios from "axios";
import Error from "../components/Error";
import { useAuthState } from "../context/auth-context";

import { ErrorResponse, BlogResponse } from "../types/types";
import { Card, FormButton } from "../theme";
import { navigate } from "@reach/router";

interface IFormInput {
  title: string;
  description: string;
}

const AddBlog = (): JSX.Element => {
  const authState = useAuthState();
  const [text, setText] = React.useState<IFormInput>({
    title: "",
    description: ""
  });
  const [showError, setShowError] = React.useState<boolean>(false);
  const [errorMessage, setErrorMessage] = React.useState<string>("");

  const onChange = (
    e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => {
    setText({ ...text, [e.target.name]: e.target.value });
  };

  const closeError = () => {
    setShowError(false);
    setErrorMessage("");
  };

  const addBlog = (e: React.FormEvent) => {
    e.preventDefault();
    axios
      .post(
        "http://localhost:5000/blog",
        { title: text.title, description: text.description },
        {
          headers: {
            authorization: authState.token
          }
        }
      )
      .then((res: BlogResponse) => {
        navigate("/profile");
      })
      .catch((err: ErrorResponse) => {
        if (err.response) {
          setShowError(true);
          setErrorMessage(err.response.data.error);
        }
      });
  };

  return (
    <Center>
      {showError && <Error message={errorMessage} closeError={closeError} />}
      <Card height="480px" width="80%">
        <Inner>
          <Header>Publish Your Blog</Header>
          <div>
            <Form onSubmit={addBlog}>
              <FormControl>
                <label htmlFor="title">Title</label>
                <input
                  type="text"
                  id="title"
                  placeholder="Title"
                  name="title"
                  value={text.title}
                  required
                  onChange={onChange}
                />
              </FormControl>
              <FormControl>
                <label htmlFor="description">Description</label>
                <textarea
                  id="description"
                  placeholder="Description"
                  name="description"
                  rows={6}
                  value={text.description}
                  required
                  onChange={onChange}
                />
              </FormControl>
              <FormControl>
                <FormButton>Publish</FormButton>
              </FormControl>
            </Form>
          </div>
        </Inner>
      </Card>
    </Center>
  );
};

const Center = styled.div`
  width: 100%;
  height: 100vh;
  display: flex;
  justify-content: center;
  align-items: center;
`;

const Inner = styled.div`
  height: 100%;
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 10px;
`;

const Header = styled.h1`
  flex-basis: 20%;
  font-size: 40px;
`;
const Form = styled.form`
  flex-basis: 80%;
  height: 100%;
  width: 100%;
`;

const FormControl = styled.div`
  display: flex;
  flex-direction: column;
  width: 100%;
  margin-bottom: 20px;

  label {
    font-size: 20px;
    margin-bottom: 10px;
    text-align: center;
  }

  input,
  textarea {
    outline-color: red;
    padding: 10px;
    width: 50vw;
    background-color: #ccc;
    border-radius: 8px;
    border: none;
    font-size: 17px;
  }
`;

export default AddBlog;
