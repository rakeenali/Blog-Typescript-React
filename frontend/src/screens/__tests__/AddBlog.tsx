import "@testing-library/jest-dom/extend-expect";
import * as React from "react";
import { render, fireEvent, wait } from "@testing-library/react";
import axios from "axios";
import { navigate } from "@reach/router";

import * as AuthContext from "../../context/auth-context";
import AddBlog from "../AddBlog";

jest.mock("@reach/router", () => {
  const navigate = jest.fn();
  return {
    navigate
  };
});

beforeEach(() => {
  jest.spyOn(axios, "post").mockImplementation((...args) => {
    console.warn("axios.post is not mocked for this call", ...args);
    return Promise.reject(new Error("This must be mocked!"));
  });
});

afterEach(() => {
  const postAxios = axios.post as jest.Mock;
  postAxios.mockClear();
});

describe("<AddBlog/>", () => {
  test("renders correctly", () => {
    const { container } = render(
      <AuthContext.AuthProvider>
        <AddBlog />
      </AuthContext.AuthProvider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test("renders the form changes correctly", () => {
    const titleText = "Some title";
    const descriptionText = "Some description not about the blog";
    const { getByPlaceholderText } = render(
      <AuthContext.AuthProvider>
        <AddBlog />
      </AuthContext.AuthProvider>
    );
    const title = getByPlaceholderText("Title") as HTMLInputElement;
    const description = getByPlaceholderText("Description") as HTMLInputElement;

    fireEvent.change(title, {
      target: { value: titleText }
    });

    fireEvent.change(description, {
      target: { value: descriptionText }
    });

    expect(title.value).toEqual(titleText);
    expect(description.value).toEqual(descriptionText);
  });

  test("can add blog successfully", async () => {
    const titleText = "Some title";
    const descriptionText = "Some description not about the blog";
    const axiosPost = axios.post as jest.Mock;

    const { getByPlaceholderText, debug, queryByText } = render(
      <AuthContext.AuthProvider>
        <AddBlog />
      </AuthContext.AuthProvider>
    );
    const title = getByPlaceholderText("Title") as HTMLInputElement;
    const description = getByPlaceholderText("Description") as HTMLInputElement;
    const publishButton = queryByText("Publish");

    fireEvent.change(title, {
      target: { value: titleText }
    });

    fireEvent.change(description, {
      target: { value: descriptionText }
    });

    axiosPost.mockResolvedValueOnce({
      status: 201,
      data: {
        id: 1
      }
    });

    publishButton && fireEvent.click(publishButton);

    await wait(
      () => {
        expect(navigate).toHaveBeenCalledTimes(1);
      },
      { timeout: 100 }
    );
  });

  test("show errors correctly", async () => {
    const titleText = "Some title";
    const axiosPost = axios.post as jest.Mock;

    const { getByPlaceholderText, queryByText } = render(
      <AuthContext.AuthProvider>
        <AddBlog />
      </AuthContext.AuthProvider>
    );
    const title = getByPlaceholderText("Title") as HTMLInputElement;
    const description = getByPlaceholderText("Description") as HTMLInputElement;
    const publishButton = queryByText("Publish");

    fireEvent.change(title, {
      target: { value: titleText }
    });

    fireEvent.change(description, {
      target: { value: "" }
    });

    axiosPost.mockRejectedValue({
      response: {
        data: {
          error: "Description field is required"
        }
      }
    });

    publishButton && fireEvent.click(publishButton);

    await wait(
      () => {
        expect(
          queryByText(/Description field is required/i)
        ).toBeInTheDocument();
      },
      { timeout: 100 }
    );
  });
});
