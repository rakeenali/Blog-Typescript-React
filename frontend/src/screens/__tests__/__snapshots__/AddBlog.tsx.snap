// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddBlog/> renders correctly 1`] = `
<div
  class="sc-ifAKCX kMXFrU"
>
  <div
    class="sc-htpNat gPXkgh"
    height="480px"
    width="80%"
  >
    <div
      class="sc-EHOje ePbsYW"
    >
      <h1
        class="sc-bZQynM cZORyu"
      >
        Publish Your Blog
      </h1>
      <div>
        <form
          class="sc-gzVnrw ieXoIT"
        >
          <div
            class="sc-htoDjs ihKokW"
          >
            <label
              for="title"
            >
              Title
            </label>
            <input
              id="title"
              name="title"
              placeholder="Title"
              required=""
              type="text"
              value=""
            />
          </div>
          <div
            class="sc-htoDjs ihKokW"
          >
            <label
              for="description"
            >
              Description
            </label>
            <textarea
              id="description"
              name="description"
              placeholder="Description"
              required=""
              rows="6"
            />
          </div>
          <div
            class="sc-htoDjs ihKokW"
          >
            <button
              class="sc-bxivhb chPYTn"
            >
              Publish
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
